File: functions/utils/businessMapper.js
// functions/utils/businessMapper.js

/**
 * Maps an adoption_centers document
 * to the new unified businesses structure
 */

function normalizeStatus(status) {
    if (!status) return "pending";
    const s = String(status).toLowerCase();
    if (["approved", "rejected", "pending"].includes(s)) {
        return s;
    }
    return "pending";
}

function mapAdoptionCenterToBusiness(oldDoc = {}) {
    const status = normalizeStatus(oldDoc.status);

    return {
        type: "adoption_center",

        ownerUid: oldDoc.ownerUid || null,

        status: status,
        isVerified:
            typeof oldDoc.isVerified === "boolean"
                ? oldDoc.isVerified
                : status === "approved",

        profile: {
            displayName: oldDoc.name || "",
            description: oldDoc.description || "",
            logoUrl: oldDoc.logoUrl || null,
            coverImageUrl: oldDoc.coverImageUrl || null,
            rating:
                typeof oldDoc.rating === "number"
                    ? oldDoc.rating
                    : 0,
            reviewCount:
                typeof oldDoc.reviewsCount === "number"
                    ? oldDoc.reviewsCount
                    : 0,
        },

        contact: {
            phone: oldDoc.phone || null,
            whatsapp: oldDoc.whatsapp || null,
            instagram: oldDoc.instagram || null,
            website: oldDoc.website || null,
            city: oldDoc.city || "",
            district: oldDoc.district || "",
            address: oldDoc.address || "",
        },

        legal: {
            taxNumber: oldDoc.taxNumber || null,
            mersisNumber: oldDoc.mersisNumber || null,
            taxDocumentUrl: oldDoc.taxDocumentUrl || null,
            signatureCircularUrl: oldDoc.signatureCircularUrl || null,
        },

        // 🔐 Audit fields
        createdAt: oldDoc.createdAt || null,
        updatedAt: null, // ⚠️ در migration خود Function با serverTimestamp ست کن
        approvedAt: oldDoc.approvedAt || null,
        rejectedAt: oldDoc.rejectedAt || null,
        rejectionReason: oldDoc.rejectionReason || null,
        resolvedBy: oldDoc.resolvedBy || null,
    };
}

module.exports = {
    mapAdoptionCenterToBusiness,
};

File: lib/services/business_mapper.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/business.dart';
import '../models/business_draft.dart';

class BusinessMapper {
  static Business fromDraft({
    required BusinessDraft draft,
    required String ownerUid,
  }) {
    return Business(
      id: '', // Firestore خودش میده
      type: "veterinary",

      ownerUid: ownerUid,
      ownerName: draft.profile.displayName,
      ownerRole: "veterinarian",

      name: draft.profile.displayName,
      description: draft.profile.description,
      yearEstablished: null,

      phone: draft.contact.phone,
      landline: null,
      email: draft.contact.email,

      address: draft.contact.addressLine,
      city: draft.contact.city,
      district: draft.contact.district,

      location: const GeoPoint(0, 0), // بعداً از map بگیر

      website: draft.contact.website,
      instagram: draft.contact.instagram,
      whatsapp: draft.contact.whatsapp,

      license: License(
        number: draft.legal.taxNumber,
        authority: "unknown",
        expiryDate: null,
        documentUrl: null,
      ),

      services: [], // بعداً پر میشه

      workingHours: {},

      emergency: Emergency(
        available: false,
        type: "none",
      ),

      animalTypes: ["dog", "cat"],

      features: Features(
        homeService: false,
        onlineConsultation: false,
        parking: false,
      ),

      logoUrl: draft.profile.logoUrl,
      images: [],

      rating: 0,
      reviewCount: 0,

      partnership: Partnership(
        model: "subscription",
        status: "pending", // 🔥 مهم
        featured: false,
      ),

      payment: Payment(
        iban: "",
        accountHolder: draft.profile.displayName,
        billingInfo: draft.legal.mersisNumber,
      ),

      marketing: Marketing(
        offersEnabled: false,
        hasDiscount: false,
        featuredScore: 0,
      ),

      isVerified: false,
      isActive: true,

      moderation: Moderation(
        status: "active",
        reportCount: 0,
      ),

      createdAt: Timestamp.now(),
      updatedAt: Timestamp.now(),
    );
  }
}